dev-docs/RFCs/module-structure-rfc.md

```diff
@@ -0,0 +1,50 @@
+# RFC: loaders.gl Module Structure
+
+* **Author**: Ib Green
+* **Date**: Dec, 2018
+* **Status**: **Draft**
+
+
+## Summary
+
+This RFC proposes an npm module structure to manage the amount of code and dependencies that applications have to include from loaders.gl.
+
+
+## Background
+
+loaders.gl is expected to grow with additional loaders, e.g. TopoJSON, more extensive KML, additional mesh and point cloud formats etc.
+
+
+## Alternatives
+
+* Placing all loaders in a single npm module creates a big, growing dependency and even with fully optimized support for tree-shaking, there will be impact on development builds, and make dependency averse developers think twice about using the library.
+
+* Splitting every loader, encoder and write into its own module is possible but easily leads to a system with dozens of modules and gets unmanageable in its own way.
+
+* Dividing into groups of related loaders ensures that apps don't have to import big chunks of code they are not even remotely intending to use, while avoiding too much book-keeping.
+
+
+## Proposal
+
+Define groups of related loaders:
+
+| Module                     | Contents  |
+| ---                        | ---       |
+| `loaders.gl`               | Umbrella module, just imports and forwards all other modules. |
+| `@loaders.gl/core`         | The user function library and the common helper functions for loaders. |
+| `@loaders.gl/mesh-loaders` | Mesh and Point-Cloud Category loaders (PLY, PCS, LAZ, OBJ) |
+| `@loaders.gl/geospatial-loaders` | Geospatial format loaders (KML, TopoJSON, ...). |
```

Comment:
Personally I think it will be a lot clearer if we just create one submodule for each format. It will also be an easy sell from the SEO or branding perspective.

Reply:
> Personally I think it will be a lot clearer if we just create one submodule for each format. It will also be an easy sell from the SEO or branding perspective.

Yes, perhaps you are right. Some observations:

* We sometimes have both loaders and writers for the same format, but splitting those into separate would be overkill. I think we can rely on tree shaking in that case.
* It will be a bunch of modules (probably dozens at some point, but that should not be an issue with our current monorepo infrastructure.

On a side note, I am also thinking right now the module should focus more on point cloud formats like PCD and LAZ rather than mesh formats like OBJ etc (we want to focus on glTF for meshes). I will look through the docs and make subtle changes to emphasize this.